### ISSUE TITLE
security: roaring_bitmap_portable_deserialize() gives undefined behavior on evil input

### ISSUE THREAD

**Issue by adamhooper** [RoaringBitmap/CRoaring#116]
I can't find a way to test that an input buffer is valid. Also, I can't specify the input buffer in `roaring_bitmap_portable_deserialize()`. It only accepts a `const char*` -- there's no accompanying `size_t`, which is _necessary_ to describe a buffer.

An evil person can craft short, evil arguments to `roaring_bitmap_portable_deserialize()` that convince CRoaring to read beyond its memory bounds and expose private data. That means I can't let an API client supply me with a RoaringBitmap to describe his or her integers.

CRoaring should not read beyond the bounds of the buffer: it should produce an error instead. The simplest way to achieve this would be:

1. Require a `size_t len` in `roaring_bitmap_portable_deserialize()`
2. Disallow reading the buffer before the beginning of `buf` or after the end of `buf+len`.
3. Make `roaring_bitmap_portable_deserialize()` return `NULL` if the input is invalid.

**Comment by lemire** [comment]
You are (obviously) correct.

**Comment by lemire** [comment]
I don't think it is hard to implement. Of course, we would need tests where you feed in garbage.

**Comment by K2** [comment]
I can defiantly help out here, was just getting some personal projects time now the holidays are soon, there's a decent amount of places if were looking to harden the API.  If we assume 32 and 64 bit widths, that adds a lot of potential around the areas where we mix int and size_t with respect to promotion and truncation that can cause out of bounds or corruption. 

Is there anything else than the serialization that makes sense to protect? We don't claim to be MT safe so that makes life a bit easier. Would any of the bindings IPC their bitmaps around in some other way, pipe or something...?   I took a quick look at the Java implementation and it seems no JNI is used, I took an imported function that [CRoaring.NET](https://github.com/RogueException/CRoaring.Net/blob/5a6d5bb176c49fb96923450bb1fe6f120694a3b6/src/CRoaring.Net/NativeMethods.cs#L13) uses as an example of what we would need to change.

We would want to look at some optimized overflow routines, determine the best placement, here's how a friend of mine suggested a possible implementation (his blog post on the subject https://www.fefe.de/intof.html).

More or less, every place we allocate memory where we do not preserve the result of an expression derived in the scope of the parameter passed to the routine doing the allocation is an instant bug and has to be fixed.  It can be argued the other way but I've got plenty of reasons why it's a bug for security and performance at the same time (since you will be required to re-calculate that value repeatedly if you're trying to use that buffer safely) (second example that would need additional checking from GO binding https://github.com/RoaringBitmap/gocroaring/blob/master/roaring.c#L2835) 

The main questions would be; 
- Do we know of other scenarios/threats to protect?
- Make this an always on sort of thing? 
       Or have it opt-in? (rpmalloc did not want the perf hit so chose to optionally define it 
       https://github.com/rampantpixels/rpmalloc/issues/26)  

Then just figure out the details of a good pattern we like for optimally handling the API, e.g. required changes to roaring_array.c

```C

int umult32(uint32_t a,uint32_t b,uint32_t* c) {
    unsigned long long x=(unsigned long long)a*b;
    if (x>0xffffffff) return 0;
    *c=x&0xffffffff;
    return 1;
  }

bool ra_init_with_capacity(roaring_array_t *new_ra, uint32_t cap) {
    if (!new_ra) return false;
    new_ra->keys = NULL;
    new_ra->containers = NULL;
    new_ra->typecodes = NULL;

    new_ra->allocation_size = cap;
    new_ra->size = 0;

    uint32_t Header = sizeof(uint16_t) + sizeof(void *) + sizeof(uint8_t);
    uint32_t neededSize;
    
    if(!umult32(cap, Header, &neededSize))
        return false;

    void *bigalloc = malloc(neededSize);

    if( bigalloc == NULL ) return false;
    new_ra->containers = (void **)bigalloc;
    new_ra->keys = (uint16_t *)(new_ra->containers + cap);
    new_ra->typecodes = (uint8_t *)(new_ra->keys + cap);
    return true;
}
```

**Comment by K2** [comment]
@lemire I took a quick look at the go code, it may need some fixes, I've done a few go reviews but there might be some platform level support on the way => https://github.com/golang/go/issues/19624
That would really help out for this sort of an issue for them.

**Comment by lemire** [comment]
@K2 Excellent! I’ll get back to you later (today?) on your specific comments.

**Comment by lemire** [comment]
@adamhooper 

I don't know how "safe" we want things to be. I have added the following functions which should go a long way toward meeting your needs:

```
/**
 * read a bitmap from a serialized version in a safe manner (reading up to maxbytes).
 * This is meant to be compatible with
 * the Java and Go versions. See format specification at
 * https://github.com/RoaringBitmap/RoaringFormatSpec
 * In case of failure, a null pointer is returned.
 */
roaring_bitmap_t *roaring_bitmap_portable_deserialize_safe(const char *buf, size_t maxbytes);

/**
 * Check how many bytes would be read (up to maxbytes) at this pointer if there
 * is a bitmap, returns zero if there is no valid bitmap.
 * This is meant to be compatible with
 * the Java and Go versions. See format specification at
 * https://github.com/RoaringBitmap/RoaringFormatSpec
 */
size_t roaring_bitmap_portable_deserialize_size(const char *buf, size_t maxbytes);
```

@K2 

My understanding is that you are raising other issues, e.g., multiplication overflows. My guess is that this is not what the OP had in mind. I suggest you open a distinct issue regarding arithmetic overflows otherwise it gets too difficult to track progress on specific issues.


I am closing this, if anyone disagrees, please re-open.

**Comment by adamhooper** [comment]
Wow, you guys are responsive :). Thanks for your great work!

What I'm hoping for from an eventual "safe" method is probably what you're thinking:

* Return `NULL` if asked to read outside buffer bounds.
* Return `NULL` on every "invalid" bitmap.

I have scant knowledge of the spec, so I don't know what an "invalid" bitmap might be. Here are things I worry about:

* A tiny amount of input leading this library to allocate an enormous amount of memory.
* A tiny amount of input leading this library to use an enormous amount of CPU. Worst case would be an infinite loop. Also bad would be excessive iteration.
* Any input leading to undefined behavior.

Basically, I'd love to use this library the way I'd use zlib: worry-free. My production website passes raw bitmaps around right now of up to 1M bits. As I understand it, roaring bitmaps could save oodles of bandwidth -- especially for the super-common cases of all-`1` and all-`0`.

**Comment by lemire** [comment]
@adamhooper 


```
* Return `NULL` if asked to read outside buffer bounds.
* Return `NULL` on every "invalid" bitmap.
```

Yep.


```
* A tiny amount of input leading this library to allocate an enormous amount of memory.
```

To a good approximation, the serialized format uses about as much memory as the "in memory" data structure. In fact, both the Go and Java have support for memory mapping, where you don't even copy the data and access it in-place. We have not yet implemented this in C... although it is not hard, especially because copy-on-write is already implemented. (We are just limited by time and the need to do things properly... but this particular feature is probably a few hours of work. Hardest part is to validate it.)

```
* A tiny amount of input leading this library to use an enormous amount of CPU. Worst case would be an infinite loop. Also bad would be excessive iteration.
```

Reading the data should be nearly free. We just copy things... as written above, in Java and Go, we even have copy-free loading.

```
* Any input leading to undefined behavior.
```

We do test for undefined behavior in the code, and if any were to be found, it would be a bug.

```
Basically, I'd love to use this library the way I'd use zlib: worry-free. My production website passes raw bitmaps around right now of up to 1M bits. As I understand it, roaring bitmaps could save oodles of bandwidth -- especially for the super-common cases of all-`1` and all-`0`.
```

Right. So the format (Roaring) is [widely used](http://roaringbitmap.org) so it is reasonable to expect that it might work. Of course, always run your own tests and benchmarks.
